linkerd/app/outbound/src/http/logical/policy/router.rs

```diff
@@ -113,58 +117,72 @@ where
     fn from((rts, parent): (Params<M, F, E>, T)) -> Self {
         let Params {
             addr,
+            meta: parent_ref,
             routes,
             backends,
             failure_accrual,
         } = rts;
 
         let mk_concrete = {
             let parent = parent.clone();
-            move |target: concrete::Dispatch| {
+            move |backend_ref: BackendRef, target: concrete::Dispatch| {
                 // XXX With policies we don't have a top-level authority name at
                 // the moment. So, instead, we use the concrete addr used for
                 // discovery for now.
                 let authority = match target {
-                    concrete::Dispatch::Balance(ref a, _) => Some(a.as_http_authority()),
+                    concrete::Dispatch::Balance(ref addr, ..) => Some(addr.as_http_authority()),
                     _ => None,
                 };
                 Concrete {
                     target,
                     authority,
                     parent: parent.clone(),
+                    backend_ref,
+                    parent_ref: parent_ref.clone(),
                     failure_accrual,
                 }
             }
         };
 
-        let mk_dispatch = move |bd: &policy::BackendDispatcher| match *bd {
+        let mk_dispatch = move |bke: &policy::Backend| match bke.dispatcher {
             policy::BackendDispatcher::BalanceP2c(
                 policy::Load::PeakEwma(policy::PeakEwma { decay, default_rtt }),
                 policy::EndpointDiscovery::DestinationGet { ref path },
-            ) => mk_concrete(concrete::Dispatch::Balance(
-                path.parse().expect("destination must be a nameaddr"),
-                http::balance::EwmaConfig { decay, default_rtt },
-            )),
-            policy::BackendDispatcher::Forward(addr, ref metadata) => mk_concrete(
-                concrete::Dispatch::Forward(Remote(ServerAddr(addr)), metadata.clone()),
+            ) => mk_concrete(
+                BackendRef(bke.meta.clone()),
+                concrete::Dispatch::Balance(
+                    path.parse::<NameAddr>()
+                        .expect("destination must be a nameaddr"),
+                    http::balance::EwmaConfig { decay, default_rtt },
+                ),
+            ),
+            policy::BackendDispatcher::Forward(addr, ref md) => mk_concrete(
+                EndpointRef::new(md, addr.port().try_into().expect("port must not be 0")).into(),
+                concrete::Dispatch::Forward(Remote(ServerAddr(addr)), md.clone()),
             ),
-            policy::BackendDispatcher::Fail { ref message } => {
-                mk_concrete(concrete::Dispatch::Fail {
+            policy::BackendDispatcher::Fail { ref message } => mk_concrete(
+                BackendRef(policy::Meta::new_default("fail")),
```

Comment:
not a blocker for now, but it occurs to me that the `fail`ing backends _could_ potentially have non-default `BackendRef` metadata, since they're returned by the controller when a backend service doesn't exist...so they should have the metadata of the referenced (non-existent) service.